pkg/policy/prefilter.go

```diff
@@ -0,0 +1,303 @@
+	var undoQueue []net.IPNet
+	var ret error
+
+	p.Mutex.Lock()
+	defer p.Mutex.Unlock()
+	if p.Revision != revision {
+		return fmt.Errorf("Latest revision is %d not %d", p.Revision, revision)
+	}
+	for _, cidr := range cidrs {
+		ones, bits := cidr.Mask.Size()
+		which := p.selectMap(ones, bits)
+		if which == mapCount || p.Maps[which] == nil {
+			ret = fmt.Errorf("No map enabled for CIDR string %s", cidr.String())
+			break
+		}
+		// Lets check obvious cases first, so we don't need to painfully unroll
+		if p.Maps[which].CIDRExists(cidr) == false {
+			ret = fmt.Errorf("No map entry for CIDR string %s", cidr.String())
+			break
+		}
+	}
+	if ret != nil {
+		return ret
+	}
+	for _, cidr := range cidrs {
+		ones, bits := cidr.Mask.Size()
+		which := p.selectMap(ones, bits)
+		err := p.Maps[which].DeleteCIDR(cidr)
+		if err != nil {
+			ret = fmt.Errorf("Error deleting CIDR string %s: %s", cidr.String(), err)
+			break
+		} else {
+			undoQueue = append(undoQueue, cidr)
+		}
+	}
+	if ret == nil {
+		p.Revision++
+		return ret
+	}
+	for _, cidr := range undoQueue {
+		ones, bits := cidr.Mask.Size()
+		which := p.selectMap(ones, bits)
+		p.Maps[which].InsertCIDR(cidr)
+	}
+	return ret
+}
+
+func (p *PreFilter) initOneMap(which preFilterMapType) error {
+	var prefixdyn bool
+	var prefixlen int
+	var maxelems uint32
+	var path string
+	var err error
+	var skip bool
+
+	switch which {
+	case prefixesV4Dyn:
+		prefixlen = 32
+		prefixdyn = true
+		maxelems = maxLKeys
+		path = bpf.MapPath(cidrmap.MapName + "v4_dyn")
+		skip = p.Config.dyn4Enabled == false
+	case prefixesV4Fix:
+		prefixlen = 32
+		prefixdyn = false
+		maxelems = maxHKeys
+		path = bpf.MapPath(cidrmap.MapName + "v4_fix")
+		skip = p.Config.fix4Enabled == false
+	case prefixesV6Dyn:
+		prefixlen = 128
+		prefixdyn = true
+		maxelems = maxLKeys
+		path = bpf.MapPath(cidrmap.MapName + "v6_dyn")
+		skip = p.Config.dyn6Enabled == false
+	case prefixesV6Fix:
+		prefixlen = 128
+		prefixdyn = false
+		maxelems = maxHKeys
+		path = bpf.MapPath(cidrmap.MapName + "v6_fix")
+		skip = p.Config.fix4Enabled == false
+	}
+	if skip == false {
+		p.Maps[which], _, err = cidrmap.OpenMapElems(path, prefixlen, prefixdyn, maxelems)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *PreFilter) init() (*PreFilter, error) {
+	var err error
+	for i := prefixesV4Dyn; i < mapCount; i++ {
+		err = p.initOneMap(i)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return p, nil
+}
+
+// ProbePreFilter checks whether XDP mode is supported on given device
+func ProbePreFilter(device, mode string) error {
+	cmd := exec.Command("ip", "-force", "link", "set", "dev", device, mode, "off")
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("Cannot run ip command: %v", err)
+	}
+	if err := cmd.Wait(); err != nil {
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+				switch status.ExitStatus() {
+				case 0:
```

Comment:
Can this happen? Exiting with an error and the exit status is zero?

Reply:
I think you're right, will test in a sec.

Reply:
Yes, fixed, thanks.